### PR TITLE
Refactor CopStore to not inherit from array and cache properties properly

### LINKED ISSUE #3827: Cop name collision with rubocop-rspec
I was playing with the current rubocop build and noticed a new warning:

```
/dev/projects/rubocop/.rubocop.yml: RSpec/FilePath has the wrong namespace - should be Rails
```

This might be confusing for users that depend on rubocop-rspec and rubocop itself. I think the proper solution here would be to not emit this warning if a cop exists for each case (`RuboCop::Cop::RSpec::FilePath` and `RuboCop::Cop::Rails::FilePath`). Before releasing the current build on master would you be open to fixing this? If you don't want this to be a release blocker it seems like it might also be valid to rename the rails cop to `Rails/RailsRoot` maybe.

### PR DESCRIPTION
**Pull request by backus**
Previously the `CopStore` class stored memoized information like cop types and cop names on first invocation. I think this caused the issues in #3827 because `cop_types` and cop name information would be memoized before rspec cops were introduced meaning that `RuboCop::Cop::Cop.all.cop_names` wouldn't include rspec cops.

The `CopStore` inherited from `Array` which meant that its behavior was smeared across many different collaborators making it hard to know when the problematic behavior was fixed. I've refactored the class to not inherit from array and pull in the behavior that collaborators were depending on through the wide array interface it was inheriting.

To fix #3827 I simply moved the cop type and cop name tracking to a set that is updated when a new cop is registered. This removes the need to memoize on first invocation without any performance loss.

`CopStore` basically had no tests so I filled them out as well so that I could be fairly confident with any refactoring it required.

Summary from commit:

 - Rename `Cop::Cop::CopStore` to `Cop::Cop::Registry`

 - Refactor CopStore to not inherit from Array. Moved functionality
   that depended on array interface into public api of CopStore

 - Improve test coverage of entire CopStore interface

 - Replaces `CopStore#<<` with `Registry#enlist` which is responsible
   for tracking `cop_names`, unique `departments`, and all cops in the
   store

   - Moving this logic to `#enlist` fixes #3827 since it eliminates the
     case where `#cop_names` can be invoked (memoizing the list of
     names) before all cops are defined

 - Extract error message building to AmbiguousCopName class

 - Update documentation that previously incorrectly said the
   method returned an array of strings

 - Removed confusing usage of `File.dirname` and `File.basename`
   for handling strings like "Style/IndentyArray". Replaced this
   with `Cop::Badge` string splitting that doesn't have edge cases like
   the previous implementation did. Using `File` to split strings
   here is especially confusing given that the same method does
   work with a file path.

   - For example, the previous implementation had a conditional
     that relied on `found_ns != File.dirname(name)` where the
     `File.dirname(name)` would be '.' when the name was not
     qualified. The new code explicitly handles this case

   - Also renamed corresponding local variables naming that seemed
     to mirror the misleading `File` module usage.

 - Adds documentation for `qualified_cop_name` since it has many edge
   cases

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests are passing.
* [x] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

### CODE REVIEW

File: lib/rubocop/cop/cop_store.rb [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  module Cop
+    # Error raised when an unqualified identifier is used that could
+    # refer to two or more cops under different namespaces
+    class AmbiguousCopName < RuboCop::Error
+      MSG = 'Ambiguous cop name `%<name>s` used in %<origin>s needs namespace' \
+            ' qualifier. Did you mean %<options>s?'.freeze
+
+      def initialize(identifier, origin, namespaces)
+        options = namespaces.map { |namespace| "#{namespace}/#{identifier}" }
+        message = MSG % {
```
Comment (bbatsov) [inline lib/rubocop/cop/cop_store.rb]: Use `format` here. I hate `%`.

Comment (bbatsov) [inline lib/rubocop/cop/cop_store.rb]: This also shows how we can improve our `%` detection cop - it should check for a hash literal as it's argument. :-)

Comment (backus) [inline lib/rubocop/cop/cop_store.rb]: :thumbsup:

Comment (backus) [inline lib/rubocop/cop/cop_store.rb]: Yeah I assumed that rubocop would catch if I used the wrong formatting style here

Comment (backus) [inline lib/rubocop/cop/cop_store.rb]: I'll open an issue for that

File: spec/rubocop/cli/cli_options_spec.rb [hunk 2]
```diff
@@ -584,7 +584,7 @@ def full_description_of_cop(cop)
       end
 
       it 'prints all available cops and their description' do
-        cops.each do |cop|
+        cops.to_a.each do |cop|
```
Comment (bbatsov) [inline spec/rubocop/cli/cli_options_spec.rb]: I definitely don't like having to invoke `to_a` all over the place.

File: lib/rubocop/runner.rb [hunk 3]
```diff
@@ -254,7 +259,7 @@ def enable_rails_cops(config)
     def mobilized_cop_classes(config)
       @mobilized_cop_classes ||= {}
       @mobilized_cop_classes[config.object_id] ||= begin
-        cop_classes = Cop::Cop.all
+        cop_classes = Cop::Cop.all.to_a
```
Comment (backus) [inline lib/rubocop/runner.rb]: I could have the `all` method be the thing that invokes `to_a` which would resolve it being called in many places.

Comment (bbatsov) [inline lib/rubocop/runner.rb]: Obviously the problem here is that this returns a `CopStore` as opposed to an array of cops. An API that's not intuitive is not a good API.

File: lib/rubocop/cop/cop_store.rb (hunk 1 above)
Comment (backus) [inline lib/rubocop/cop/cop_store.rb]: See https://github.com/bbatsov/rubocop/issues/3839

**Comment by bbatsov** [comment]
I welcome the changes overall, but I've added a few minor remarks inline.

**Comment by bbatsov** [comment]
And I think we've strayed from our fun police references! `cop_type`? A cop should have a department, not a type! :-)

**Comment by bbatsov** [comment]
I'd even name CopStore `Precinct` instead, but I guess not everyone would appreciate my sense of humour. :D

**Comment by backus** [comment]
@bbatsov addressed most of your flags. Reduced `to_a` calls by also exposing `Cop::Cop.cop_store` and using that or `Cop::Cop.all` depending on the context.

File: lib/rubocop/cli.rb [hunk 4]
```diff
@@ -104,7 +104,7 @@ def apply_default_formatter
     def print_available_cops
       # Load the configs so the require()s are done for custom cops
       @config_store.for(Dir.pwd)
-      cops = Cop::Cop.all
+      cops = Cop::Cop.cop_store
```
Comment (bbatsov) [inline lib/rubocop/cli.rb]: This looks a bit odd - I would not expect something named cops to be an instance of `cop_store`.

File: lib/rubocop/cli.rb [hunk 5]
```diff
@@ -140,7 +140,7 @@ def selected_cops_of_type(cops, type)
     end
 
     def cops_of_type(cops, type)
-      cops.with_type(type).sort_by!(&:cop_name)
+      cops.with_type(type).to_a.sort_by!(&:cop_name)
```
Comment (bbatsov) [inline lib/rubocop/cli.rb]: Same here.

File: lib/rubocop/cop/cop_store.rb [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+          origin: origin,
+          options: options.join(' or ')
+        )
+
+        super(message)
+      end
+    end
+
+    # Store for all cops with helper functions
+    class CopStore
```
Comment (bbatsov) [inline lib/rubocop/cop/cop_store.rb]: Btw, I really dislike this name. I think `CopRegistry` or something similar would be way more informative.

File: lib/rubocop/cop/cop_store.rb [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        @cop_names = Set.new
+        @cop_types = {}
+
+        cops.each { |cop| register(cop) }
+      end
+
+      def register(cop)
+        @cops << cop
+        @cop_names.add(cop.cop_name)
+        @cop_types[cop.cop_type] ||= namespace(cop.cop_name)
```
Comment (bbatsov) [inline lib/rubocop/cop/cop_store.rb]: We use namespace and type interchangeably it would seem. I'd just go with the term `department`.

**Comment by bbatsov** [comment]
I went over the code one more time and added a few more remarks. In general the entire `CopStore` class could benefits from some cleanup/terminology unification. A few things there seem confusing to me and I'm reasonably sure I didn't author them. :-) In particular it seems that some functionality there doesn't pertain strictly to a cop registry and some concepts are completed - types and namespaces for instance.

File: lib/rubocop/cop/cop_store.rb (hunk 6 above)
Comment (backus) [inline lib/rubocop/cop/cop_store.rb]: Sure I can change that. I didn't before since I didn't introduce `CopStore`

Comment (bbatsov) [inline lib/rubocop/cop/cop_store.rb]: I know. :-) Guess I just didn't pay much attention to the name when this was originally introduced.

**Comment by backus** [comment]
@bbatsov I think this is finally ready for another pass. @Drenmi I'd also be interested in your input here if you're free

**Comment by bbatsov** [comment]
Apart from the failing build - the changes look good to me. As soon as Travis is green I'll merge the PR.

**Comment by backus** [comment]
@bbatsov Thanks. Build is green now!

**Comment by bbatsov** [comment]
👍

**Comment by NickLaMuro** [comment]
@backus @bbatsov Hi there.

Noticed that a CHANGELOG entry is absent for this PR (even though it is "checked" as being so in the description).  With the addition of the `RuboCop::Cop::Registry`, `RuboCop::Cop::Team.new` no longer supports passing in an array of `Cops`, but requires that they be sent in with the registry wrapper.

This is a fine change, but a breaking one for any third party libs (like `rubocop-git`, which I am currently trying to help support) which use the underlying API in it's code base.  Would hope that you would retroactively add this to the CHANGELOG for the 0.47.0 release so it is properly documented for those in the future.

Thanks!

**Comment by bbatsov** [comment]
> This is a fine change, but a breaking one for any third party libs (like rubocop-git, which I am currently trying to help support) which use the underlying API in it's code base. Would hope that you would retroactively add this to the CHANGELOG for the 0.47.0 release so it is properly documented for those in the future.

We don't really have a formal public API yet, so we don't log such changes in the changelog. There are a few open tickets regarding this, but we haven't made much progress on them recently.

**Comment by backus** [comment]
@NickLaMuro I can possibly help with that at some point. I made this change and also maintain rubocop-rspec

**Comment by NickLaMuro** [comment]
> We don't really have a formal public API yet, so we don't log such changes in the changelog. There are a few open tickets regarding this, but we haven't made much progress on them recently.

@bbatsov That is fair, and I forget this is technically "beta" even though it is very widely used.  Again, no problem with the change, just figure it wouldn't have minded a mention in the "CHANGES" section of the changelog, but I understand the decision.  Thanks!

> @NickLaMuro I can possibly help with that at some point. I made this change and also maintain rubocop-rspec

@backus Actually, I already have that covered in my second commit to https://github.com/m4i/rubocop-git/pull/26 😄 

I, by fluke, landed on this PR (first try!) when digging through the changes to `RuboCop::Cop::Team`, trying to figure out what was causing the issue on newer versions, and already had a solution in mind for `rubocop-git` before posting here.  Main intent was to make sure that a "crazy git history dive" wasn't required by another developer in the future when such a well maintained CHANGELOG exists (something I really admire about this project), but again, I understand the rational for not including this PR there.

Regardless, I appreciate the offer to help, thanks!

**Comment by backus** [comment]
:thumbsup:. Regardless, here are a few issues I want to eventually get to that you might be interested in @NickLaMuro:

* https://github.com/bbatsov/rubocop/issues/3414 - Formalize public API for registering 3rd party cops (@drenmi is also working on this I think)

* https://github.com/backus/rubocop-rspec/issues/246 - Extract part of project to rubocop-extension. Possibly could lessen pain of maintaining third party rubocop extensions soon

**Pull request merged**